// angular2-image-upload/lib/image-upload.module.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var core_1 = require("@angular/core");
var image_upload_component_1 = require("./image-upload/image-upload.component");
var file_drop_directive_1 = require("./file-drop.directive");
var common_1 = require("@angular/common");
var image_service_1 = require("./image.service");
var ImageUploadModule = ImageUploadModule_1 = (function () {
    function ImageUploadModule() {
    }
    ImageUploadModule.forRoot = function () {
        return {
            ngModule: ImageUploadModule_1,
            providers: [image_service_1.ImageService]
        };
    };
    return ImageUploadModule;
}());
ImageUploadModule = ImageUploadModule_1 = __decorate([
    core_1.NgModule({
        imports: [common_1.CommonModule],
        declarations: [
            image_upload_component_1.ImageUploadComponent,
            file_drop_directive_1.FileDropDirective
        ],
        exports: [image_upload_component_1.ImageUploadComponent]
    })
], ImageUploadModule);
exports.ImageUploadModule = ImageUploadModule;
var ImageUploadModule_1;
